File: vtp-pensjon-application/src/main/java/no/nav/pensjon/vtp/application/configuration/SoapWebServiceConfig.java
package no.nav.pensjon.vtp.application.configuration;

import static java.lang.String.join;
import static java.util.Arrays.stream;
import static java.util.Optional.ofNullable;

import static org.slf4j.LoggerFactory.getLogger;

import java.lang.reflect.Method;
import java.util.List;
import java.util.Optional;

import javax.annotation.PostConstruct;
import javax.servlet.http.HttpServletResponse;

import org.apache.cxf.Bus;
import org.apache.cxf.jaxws.EndpointImpl;
import org.apache.cxf.jaxws.JAXWSMethodInvoker;
import org.apache.cxf.jaxws.JaxWsServerFactoryBean;
import org.apache.cxf.message.Exchange;
import org.apache.cxf.service.invoker.Factory;
import org.apache.cxf.service.invoker.Invoker;
import org.apache.cxf.service.invoker.SingletonFactory;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import no.nav.pensjon.vtp.core.annotations.SoapService;

@Configuration
public class SoapWebServiceConfig {
    private final ApplicationContext applicationContext;
    private final Bus bus;

    public SoapWebServiceConfig(ApplicationContext applicationContext, Bus bus) {
        this.applicationContext = applicationContext;
        this.bus = bus;
    }

    @PostConstruct
    public void registerSoapServices() {
        applicationContext.getBeansWithAnnotation(SoapService.class)
                .values()
                .stream().parallel()
                .forEach(service ->
                        stream(ofNullable(service.getClass().getAnnotation(SoapService.class))
                                .map(SoapService::path)
                                .orElseThrow(() -> new RuntimeException("Missing @" + SoapService.class.getSimpleName() + " on class " + service.getClass().getSimpleName())))
                                .forEach(path -> new EndpointImpl(bus, service, new JaxWsMethodInspectingInvokerServerFactoryBean()).publish(path))
                );
    }

    private static class JaxWsMethodInspectingInvokerServerFactoryBean extends JaxWsServerFactoryBean {
        @Override
        protected Invoker createInvoker() {
            if (getServiceBean() == null) {
                return new JAXWSMethodInspectingInvoker(new SingletonFactory(getServiceClass()));
            }
            return new JAXWSMethodInspectingInvoker(getServiceBean());
        }
    }

    /**
     * Adds the x-vtp-handler to make it easy to figure which java method
     * that was invoked. Logs any exception that occurs during invocation
     * for easier development and debugging.
     */
    private static class JAXWSMethodInspectingInvoker extends JAXWSMethodInvoker {
        private final Logger logger = getLogger(getClass());

        public JAXWSMethodInspectingInvoker(Object bean) {
            super(bean);
        }

        public JAXWSMethodInspectingInvoker(Factory factory) {
            super(factory);
        }

        @Override
        protected Object invoke(Exchange exchange, Object serviceObject, Method m, List<Object> params) {
            getHttpServletResponse()
                    .ifPresent(response -> response.addHeader("x-vtp-handler", formatHandlerString(m)));

            try {
                return super.invoke(exchange, serviceObject, m, params);
            } catch (Exception e) {
                logger.error("Error invoking soap web service", e);
                throw e;
            }
        }

        @NotNull
        private String formatHandlerString(Method method) {
            final String classCanonicalName = method.getDeclaringClass().getCanonicalName();
            final String methodName = method.getName();
            final String parameters = join(",", (Iterable<String>) stream(method.getParameterTypes()).map(Class::getSimpleName)::iterator);
            return classCanonicalName + "#" + methodName + "(" + parameters + ")";
        }

        @NotNull
        private Optional<HttpServletResponse> getHttpServletResponse() {
            return ofNullable(RequestContextHolder.getRequestAttributes())
                    .filter(ServletRequestAttributes.class::isInstance)
                    .map(ServletRequestAttributes.class::cast)
                    .map(ServletRequestAttributes::getResponse);
        }
    }
}

File: vtp-pensjon-application/src/main/java/no/nav/pensjon/vtp/application/configuration/SoapWebServiceConfig.kt
package no.nav.pensjon.vtp.application.configuration

import org.apache.cxf.Bus
import javax.annotation.PostConstruct
import no.nav.pensjon.vtp.core.annotations.SoapService
import no.nav.pensjon.vtp.utilities.description
import org.apache.cxf.jaxws.EndpointImpl
import org.apache.cxf.jaxws.JAXWSMethodInvoker
import org.apache.cxf.jaxws.JaxWsServerFactoryBean
import org.apache.cxf.message.Exchange
import org.apache.cxf.service.invoker.Factory
import org.apache.cxf.service.invoker.Invoker
import org.apache.cxf.service.invoker.SingletonFactory
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationContext
import org.springframework.context.annotation.Configuration
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import java.lang.Exception
import java.lang.reflect.Method

@Configuration
class SoapWebServiceConfig(private val applicationContext: ApplicationContext, private val bus: Bus) {
    @PostConstruct
    fun registerSoapServices() {
        applicationContext.getBeansWithAnnotation(SoapService::class.java)
                .values
                .parallelStream()
                .forEach { service ->
                    service.javaClass.getAnnotation(SoapService::class.java)
                            ?.path?.forEach { EndpointImpl(bus, service, JaxWsMethodInspectingInvokerServerFactoryBean()).publish(it) }
                }
    }
}

class JaxWsMethodInspectingInvokerServerFactoryBean : JaxWsServerFactoryBean() {
    override fun createInvoker(): Invoker {
        return if (serviceBean == null) {
            JAXWSMethodInspectingInvoker(SingletonFactory(serviceClass))
        } else JAXWSMethodInspectingInvoker(SingletonFactory(serviceBean))
    }
}

/**
 * Adds the x-vtp-handler to make it easy to figure which java method
 * that was invoked. Logs any exception that occurs during invocation
 * for easier development and debugging.
 */
class JAXWSMethodInspectingInvoker(factory: Factory) : JAXWSMethodInvoker(factory) {
    private val logger = LoggerFactory.getLogger(javaClass)

    override fun invoke(exchange: Exchange, serviceObject: Any, method: Method, params: List<Any>): Any {
        val requestAttributes = RequestContextHolder.getRequestAttributes()
        if (requestAttributes is ServletRequestAttributes) {
            requestAttributes.response?.addHeader("x-vtp-handler", description(method))
        }

        return try {
            super.invoke(exchange, serviceObject, method, params)
        } catch (e: Exception) {
            logger.error("Error invoking soap web service", e)
            throw e
        }
    }
}
